File: scripts/verify-play-version-code.py
#!/usr/bin/env python3
"""Fail unless a Play store version code already exists as an uploaded artifact.

`make promote-android` releases an existing build to alpha/beta by writing the
target track with `--version_codes_to_retain`, without uploading anything. If
the code was never uploaded, Play rejects the track update with a generic
error that says nothing about where the code was supposed to come from. Check
it up front so the failure names the code and the run that should have
produced it.

This talks to the Play Developer API directly rather than through fastlane:
the check has to run before supply opens its own edit, and supply exposes no
read-only entry point for the list of uploaded bundles. Authentication uses
the same service-account JSON supply uses, and the RS256 assertion is signed
with the openssl CLI, so nothing beyond python3 is required.
"""

import argparse
import base64
import json
import os
import subprocess
import sys
import tempfile
import time
import urllib.error
import urllib.parse
import urllib.request

TOKEN_URL = "https://oauth2.googleapis.com/token"
SCOPE = "https://www.googleapis.com/auth/androidpublisher"
API = "https://androidpublisher.googleapis.com/androidpublisher/v3/applications"


def fail(message):
    print(f"Error: {message}", file=sys.stderr)
    sys.exit(1)


def b64url(raw):
    return base64.urlsafe_b64encode(raw).rstrip(b"=").decode("ascii")


def sign_rs256(payload, private_key_pem):
    """Sign with the service account key. openssl needs the key in a file."""
    fd, path = tempfile.mkstemp(suffix=".pem")
    try:
        os.write(fd, private_key_pem.encode("utf-8"))
        os.close(fd)
        result = subprocess.run(
            ["openssl", "dgst", "-sha256", "-sign", path, "-binary"],
            input=payload,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
        )
    except FileNotFoundError:
        fail("openssl is not on PATH; it is required to sign the Play API token")
    finally:
        os.unlink(path)

    if result.returncode != 0:
        fail(f"openssl could not sign the Play API token: {result.stderr.decode().strip()}")
    return result.stdout


def request(url, token=None, method="GET", data=None):
    headers = {}
    if token:
        headers["Authorization"] = f"Bearer {token}"
    if data is not None:
        headers["Content-Type"] = "application/x-www-form-urlencoded"
    req = urllib.request.Request(url, data=data, headers=headers, method=method)
    try:
        with urllib.request.urlopen(req, timeout=60) as response:
            body = response.read()
    except urllib.error.HTTPError as exc:
        detail = exc.read().decode("utf-8", "replace").strip()
        fail(f"Play API {method} {url} failed with HTTP {exc.code}: {detail}")
    except urllib.error.URLError as exc:
        fail(f"Play API {method} {url} could not be reached: {exc.reason}")
    return json.loads(body) if body else {}


def access_token(key):
    now = int(time.time())
    audience = key.get("token_uri", TOKEN_URL)
    header = b64url(json.dumps({"alg": "RS256", "typ": "JWT"}).encode())
    claims = b64url(
        json.dumps(
            {
                "iss": key["client_email"],
                "scope": SCOPE,
                "aud": audience,
                "iat": now,
                "exp": now + 3600,
            }
        ).encode()
    )
    signing_input = f"{header}.{claims}".encode("ascii")
    assertion = f"{header}.{claims}.{b64url(sign_rs256(signing_input, key['private_key']))}"
    body = urllib.parse.urlencode(
        {
            "grant_type": "urn:ietf:params:oauth:grant-type:jwt-bearer",
            "assertion": assertion,
        }
    ).encode()
    return request(audience, method="POST", data=body)["access_token"]


def uploaded_version_codes(package_name, token):
    """Every version code Play has an artifact for, bundles and apks alike."""
    edit = request(f"{API}/{package_name}/edits", token=token, method="POST", data=b"")
    edit_id = edit["id"]
    try:
        codes = set()
        for kind, field in (("bundles", "bundles"), ("apks", "apks")):
            listing = request(f"{API}/{package_name}/edits/{edit_id}/{kind}", token=token)
            for artifact in listing.get(field) or []:
                codes.add(int(artifact["versionCode"]))
        return codes
    finally:
        # Read-only check: never leave the throwaway edit hanging around.
        try:
            request(f"{API}/{package_name}/edits/{edit_id}", token=token, method="DELETE")
        except SystemExit:
            pass


def main():
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--json-key", required=True, help="Play service account JSON")
    parser.add_argument("--package-name", required=True)
    parser.add_argument(
        "--version-code",
        required=True,
        type=int,
        help="Store version code (VERSION_CODE_OFFSET + raw build number)",
    )
    args = parser.parse_args()

    try:
        with open(args.json_key, encoding="utf-8") as handle:
            key = json.load(handle)
    except OSError as exc:
        fail(f"could not read the Play service account key {args.json_key}: {exc}")
    except json.JSONDecodeError as exc:
        fail(f"the Play service account key {args.json_key} is not valid JSON: {exc}")

    for field in ("client_email", "private_key"):
        if not key.get(field):
            fail(f"the Play service account key {args.json_key} has no '{field}'")

    codes = uploaded_version_codes(args.package_name, access_token(key))
    if args.version_code not in codes:
        newest = max(codes) if codes else "none"
        fail(
            f"store version code {args.version_code} has never been uploaded to "
            f"{args.package_name}, so there is nothing to release.\n"
            f"       A build only becomes releasable once a ci-release run has uploaded "
            f"it to the internal track.\n"
            f"       Check that the ci-release run for this build number finished "
            f"successfully.\n"
            f"       Highest uploaded code for this package: {newest}."
        )

    print(f"store version code {args.version_code} is uploaded to {args.package_name}")


if __name__ == "__main__":
    main()

File: scripts/verify-play-version-code.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

# Fail unless a Play store version code already exists as an uploaded artifact.
#
# `make promote-android` releases an existing build to alpha/beta by writing
# the target track with `--version_codes_to_retain`, without uploading
# anything. If the code was never uploaded, Play rejects the track update with
# a generic error that says nothing about where the code was supposed to come
# from. Check it up front so the failure names the code and the run that should
# have produced it.
#
# This uses fastlane's own Play client -- Supply::Client#aab_version_codes and
# #apks_version_codes (supply/lib/supply/client.rb:268-283) are exactly this
# read, and Supply::Client.make_from_config (client.rb:16-20) does the
# service-account auth. The check has to finish before supply opens its own
# edit, which is why it opens and deletes a throwaway edit of its own rather
# than sharing one.
#
# Fails closed: every error path exits non-zero.

require 'optparse'

begin
  require 'supply'
rescue LoadError => e
  warn("Error: could not load fastlane's supply library (#{e.message}).")
  warn("       This script runs under `ruby` from PATH, which is a different install")
  warn("       from the `fastlane` command `make promote-android` calls -- Homebrew's")
  warn("       fastlane formula wraps a private GEM_HOME and does not satisfy this.")
  warn("       Install the gem for this interpreter: gem install fastlane")
  exit(1)
end

def fail_with(message)
  warn("Error: #{message}")
  exit(1)
end

options = {}
parser = OptionParser.new do |opts|
  opts.banner = "Usage: verify-play-version-code.rb --json-key KEY --package-name PKG --version-code CODE"
  opts.on("--json-key KEY", "Play service account JSON") { |v| options[:json_key] = v }
  opts.on("--package-name PKG", "Play package name") { |v| options[:package_name] = v }
  opts.on("--version-code CODE", Integer,
          "Store version code (VERSION_CODE_OFFSET + raw build number)") { |v| options[:version_code] = v }
end

begin
  parser.parse!
rescue OptionParser::ParseError => e
  fail_with("#{e.message}\n#{parser}")
end

%i[json_key package_name version_code].each do |key|
  fail_with("--#{key.to_s.tr('_', '-')} is required\n#{parser}") if options[key].nil?
end

package_name = options[:package_name]
version_code = options[:version_code]

# Build a real supply configuration rather than a bare hash, so the client gets
# supply's own defaults (notably :timeout) and its json_key verify_block, which
# checks the file exists and is JSON before any network call.
begin
  Supply.config = FastlaneCore::Configuration.create(
    Supply::Options.available_options,
    {
      json_key: options[:json_key],
      package_name: package_name
    }
  )
rescue StandardError => e
  fail_with("could not configure the Play client: #{e.message}")
end

# Each stage gets its own rescue so the message names the stage that actually
# failed. Wrapping all three together reports an auth rejection as "could not
# list uploaded artifacts", which points at the wrong thing.
begin
  client = Supply::Client.make_from_config(params: Supply.config)
rescue StandardError => e
  fail_with("could not authenticate with Play for #{package_name}: #{e.message}")
end

begin
  client.begin_edit(package_name: package_name)
rescue StandardError => e
  fail_with("could not open a Play edit for #{package_name}: #{e.message}")
end

codes = nil
begin
  codes = (client.aab_version_codes + client.apks_version_codes).map(&:to_i)
rescue StandardError => e
  fail_with("could not list uploaded artifacts for #{package_name}: #{e.message}")
ensure
  # Read-only check: never leave the throwaway edit hanging around. A failure to
  # clean it up must not mask the result of the check itself. This still runs
  # when fail_with above raises SystemExit.
  begin
    client.abort_current_edit
  rescue StandardError => e
    warn("Warning: could not delete the throwaway Play edit: #{e.message}")
  end
end

unless codes.include?(version_code)
  newest = codes.empty? ? "none" : codes.max
  fail_with(
    "store version code #{version_code} has never been uploaded to #{package_name}, " \
    "so there is nothing to release.\n" \
    "       A build only becomes releasable once a ci-release run has uploaded " \
    "it to the internal track.\n" \
    "       Check that the ci-release run for this build number finished " \
    "successfully.\n" \
    "       Highest uploaded code for this package: #{newest}."
  )
end

puts("store version code #{version_code} is uploaded to #{package_name}")
